{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "de0a7a0d-3b7d-43e9-ac5c-d45a5c610c94",
   "metadata": {},
   "source": [
    "# Differential expression analysis\n",
    "Data are EP1NS cells (ependymoma ZFTA-fusion cell line) treated with shL1CAM, shSHTN1, or shControl.\n",
    "[DESeq2 docs](https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html).\n",
    "\n",
    "## Required input files:\n",
    "*shL1CAM-shSHTN1.expected_counts.tsv*.  \n",
    "Generated by running RSEM on all samples using the following parameters:  \n",
    "`rsem-calculate-expression --star --star-gzipped-read-file --paired-end --append-names $R1 $R2 $REF $SAMPLE`  \n",
    "where the reference is ensembl 113 (GRCH38.p14)\n",
    "\n",
    "*sample_metadata.tsv*.  \n",
    "Sample metadata indicating shRNA treatments. Formatted as tsv of sample, treatment.\n",
    "\n",
    "*Homo_sapiens.GRCh38.113.gtf*.  \n",
    "Gene annotation file used for STAR, RSEM. Download from https://ftp.ensembl.org/pub/release-113/gtf/homo_sapiens/Homo_sapiens.GRCh38.113.gtf.gz."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beb7ba5d-8c9a-4b18-848a-8c48489e9a75",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Load dependencies\n",
    "\n",
    "Sys.setenv(LANGUAGE = \"en\") # set language to \"ja\" if you prefer\n",
    "suppressWarnings(library(DESeq2))\n",
    "suppressWarnings(library(plyranges))\n",
    "suppressWarnings(library(dplyr))\n",
    "suppressWarnings(library(tibble))\n",
    "suppressWarnings(library(readr))\n",
    "suppressWarnings(library(ggplot2))\n",
    "suppressWarnings(library(ggrepel))\n",
    "suppressWarnings(library(extrafont))\n",
    "suppressWarnings(library(svglite))\n",
    "suppressWarnings(library(patchwork)) # combine plots\n",
    "\n",
    "suppressMessages(extrafont::font_import(pattern=\"Arial\",prompt=FALSE))\n",
    "suppressMessages(extrafont::loadfonts())\n",
    "\n",
    "getwd()\n",
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "825e6dc9-3be2-4d36-8e03-30490ea7b017",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting defaults\n",
    "base_theme <- theme_classic(base_size=14, base_family=\"Arial\",) +\n",
    "    theme(axis.text = element_text(size=14,colour=\"black\"),\n",
    "          aspect.ratio=1,\n",
    "          #plot.margin=unit(c(0,0,0,0), \"null\")\n",
    "         )\n",
    "theme_set(base_theme)\n",
    "\n",
    "write_plot <- function(plt,outfile,width,height){\n",
    "    b=basename(outfile)\n",
    "    d=dirname(outfile)\n",
    "    dir.create(d, recursive=TRUE, showWarnings = FALSE)\n",
    "    pdf.options(encoding='ISOLatin2.enc')\n",
    "    #pdfName = paste(outfile, \".pdf\", sep=\"\")\n",
    "    pngName = paste(b, \".png\", sep=\"\")\n",
    "    svgName = paste(b, \".svg\", sep = \"\")\n",
    "    #ggsave(path=\"figures\", filename=pdfName, device=\"pdf\", width=width, height=height, units='in')\n",
    "    ggsave(path=d, device=\"png\", filename=pngName, width=width, height=height, units='in')\n",
    "    ggsave(path=d, device=\"svg\", filename=svgName, width=width, height=height, units='in')\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff8cadca-0c2b-4863-aaba-ca78380379f4",
   "metadata": {},
   "source": [
    "# Get and process your data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f49f432-7a83-45a8-9ed5-e405d16ee94b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_location = 'results/rsem/shL1CAM-shSHTN1.expected_counts.tsv'\n",
    "metadata_location = 'anno/sample_metadata.tsv'\n",
    "annotation_location = 'anno/ensembl-113/Homo_sapiens.GRCh38.113.gtf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f934944c-b482-45e6-b6ff-e867a0d1587c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "load_annotations <- function(annotation_path) {\n",
    "    # Load  gene annotations from a .gtf file.\n",
    "    # I'm using the basic primary gene annotation file from ensembl 113:\n",
    "    # https://ftp.ensembl.org/pub/release-113/gtf/homo_sapiens/Homo_sapiens.GRCh38.113.gtf.gz\n",
    "    g <- rtracklayer::import(annotation_path) %>%\n",
    "        filter(type=='gene') %>%\n",
    "       filter(gene_biotype=='protein_coding')\n",
    "    return(g)\n",
    "}\n",
    "\n",
    "filter_protein_coding <- function(gene_matrix,annotation_path){\n",
    "    # given a matrix with rownames of the form ENSG00000000003_TSPAN6,\n",
    "    # return a subset consisting of only rows where the stable gene ID ENSG00000000003 \n",
    "    # is annotated as a protein-coding gene.\n",
    "    message('filtering for protein-coding genes ...')\n",
    "    g <- load_annotations(annotation_path)\n",
    "    # need to match on stable ENSG IDs\n",
    "    ensembl_ids <- sub(\"_.*\", \"\", rownames(gene_matrix))\n",
    "    keep <- ensembl_ids %in% g$gene_id\n",
    "    filtered_data <- gene_matrix[keep, ]\n",
    "    return(filtered_data)\n",
    "}\n",
    "\n",
    "load_gex  <- function(data_location,annotation_location){\n",
    "    message(\"loading gene expression data from \", data_location, \"...\")\n",
    "    cts = as.matrix(read.csv(data_location,sep='\\t',row.names=\"gene_id\",check.names=FALSE))\n",
    "    cts = round(cts) # counts must be integer\n",
    "    cts = filter_protein_coding(cts,annotation_location)\n",
    "    new_rownames <- sub(\"^ENSG\\\\d*_\", \"\", rownames(cts)) # Remove ENSG prefixes\n",
    "    rownames(cts) <- new_rownames\n",
    "    return(cts)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f04630d4-36db-439a-b8ff-eb5639de6264",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read and preprocess expression data\n",
    "data = load_gex(data_location,annotation_location)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "361d7664-4b07-47eb-839d-3440075a94bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read and format our annotation table\n",
    "annot = read.table(metadata_location, row.names=1, header=TRUE)\n",
    "annot <- annot[match(colnames(data), rownames(annot)), , drop=FALSE]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d14cb19a-2f24-4396-aea0-e59de1f49360",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop possibly problematic controls\n",
    "# drop = c(\"shNC_REP3\",\"shNC_REP4\")\n",
    "# data = data[, setdiff(colnames(data),drop)]\n",
    "# annot = annot[setdiff(rownames(annot),drop), , drop=FALSE]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0819ece8-8904-4b42-9f21-2ff4a7738add",
   "metadata": {},
   "outputs": [],
   "source": [
    "data %>% head\n",
    "annot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "196741e2-019b-4f2f-ba26-ca66fe9bb05f",
   "metadata": {},
   "source": [
    "# Full linear model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb1c78ff-5a65-4d57-8b85-a320e60d5213",
   "metadata": {},
   "outputs": [],
   "source": [
    "dds <- DESeq2::DESeqDataSetFromMatrix(countData = data,\n",
    "                              colData = annot,\n",
    "                              design = ~ treatment)\n",
    "# Pre-filter\n",
    "smallestGroupSize <- 2\n",
    "keep <- rowSums(counts(dds) >= 10) >= smallestGroupSize\n",
    "dds <- dds[keep,]\n",
    "# Run regression\n",
    "dds <- DESeq(dds)\n",
    "# Print model summary\n",
    "dds\n",
    "# print model coefficient names\n",
    "resultsNames(dds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2770d3bd-91a7-4458-903b-6c0d4cc1a1f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Looks like much more RNA was extracted from all controls than all treatment samples.\n",
    "sizeFactors(dds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29666e82-1c3f-4709-84e2-380a7df501e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A bit of QC - looks like PC1 separates controls from treatments, good. PC2 separates shL1CAM from shSHTN1, and also good controls from slight outlier controls.\n",
    "vsd <- vst(dds, blind=TRUE)\n",
    "\n",
    "a = plotPCA(vsd, intgroup=c(\"treatment\")) + geom_label_repel(aes(label = name))\n",
    "w=6;h=6\n",
    "write_plot(a,outfile='results/deseq2/pca',width=w,height=h)\n",
    "options(repr.plot.width=6, repr.plot.height=6)\n",
    "a\n",
    "\n",
    "## Uncomment to show pairwise plots\n",
    "#vsd_s1 <- vsd[ , vsd$treatment %in% c(\"shSHTN1\", \"control\") ]\n",
    "#b = plotPCA(vsd_s1, intgroup=c(\"treatment\")) + geom_label_repel(aes(label = name))\n",
    "#vsd_s2 <- vsd[ , vsd$treatment %in% c(\"shL1CAM\", \"control\") ]\n",
    "#c = plotPCA(vsd_s2, intgroup=c(\"treatment\")) + geom_label_repel(aes(label = name))\n",
    "#vsd_s3 <- vsd[ , vsd$treatment %in% c(\"shL1CAM\", \"shSHTN1\") ]\n",
    "#d = plotPCA(vsd_s3, intgroup=c(\"treatment\")) + geom_label_repel(aes(label = name))\n",
    "# options(repr.plot.width=12, repr.plot.height=12)\n",
    "#(a | b) / (c | d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6b49baf-b65f-4209-8893-20af97d41c3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Generate DE comparisons. Wald test with BH correction. Post-hoc LFC shrinkage with apeglm.\n",
    "res_l1cam <- results(dds, name=\"treatment_shL1CAM_vs_control\",independentFiltering=TRUE)\n",
    "res_l1cam <- lfcShrink(dds, coef=\"treatment_shL1CAM_vs_control\", res=res_l1cam, type=\"apeglm\")\n",
    "\n",
    "res_shtn1 <- results(dds, name=\"treatment_shSHTN1_vs_control\",independentFiltering=TRUE)\n",
    "res_shtn1 <- lfcShrink(dds, coef=\"treatment_shSHTN1_vs_control\", res=res_shtn1, type=\"apeglm\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7ea3662-981d-49a4-8542-e0205c9f6d8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_hits <- function(deseq_result,decreasing=FALSE,outfile=NULL){\n",
    "    res_df <- as.data.frame(deseq_result)  # Convert DESeqResults object to a data frame\n",
    "    res_df <- res_df[order(res_df$log2FoldChange, decreasing=decreasing),]  # Sort by foldChange\n",
    "    # Filter by padj < 0.05\n",
    "    filtered_sorted_res <- res_df[(!is.na(res_df$padj)) & (res_df$padj < 0.05), ]  # Filter by padj\n",
    "    \n",
    "    if (!is.null(outfile)) {\n",
    "        # Write the data frame to a tab-separated file (TSV)\n",
    "        write.table(res_df, file = outfile, sep = \"\\t\", quote = FALSE)\n",
    "    }\n",
    "    return(filtered_sorted_res)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e218b24a-593c-4e25-9045-f5535f4170f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genes down in shL1CAM  \n",
    "get_hits(res_l1cam,FALSE,outfile='results/deseq2/shL1CAM_deg.tsv') %>% head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b2de680-9252-4f23-8814-e2efc6ff172e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genes down in shSHTN1  \n",
    "get_hits(res_shtn1,FALSE,outfile='results/deseq2/shSHTN1_deg.tsv') %>% head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a48497b-94c9-41cd-bde3-451b45146bf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lookup genes of interest\n",
    "genes_of_interest = c('SHTN1','L1CAM','CCN2','CCN1','YAP1') # CTGF = CCN2; CYR61 =  CCN1\n",
    "res_shtn1[rownames(res_shtn1) %in% genes_of_interest,]\n",
    "message(\"\")\n",
    "res_l1cam[rownames(res_l1cam) %in% genes_of_interest,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b6b1fab-e31c-44e9-ad28-5c3e337e1942",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Log normalized expression scaled by library size.\n",
    "# https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#extracting-transformed-values\n",
    "assay(vsd)[rownames(vsd) %in% genes_of_interest, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acd75cd3-bfa4-4fcc-b4d8-c1e691a24a6e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "fosl2-de",
   "language": "R",
   "name": "fosl2-de"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
